File: meriq-backend/routers/assessment.py
from fastapi import APIRouter
from typing import Dict, Any
from services.assessment_service import assessment_service
from models.assessment import AssessmentSubmission

router = APIRouter(prefix="/assessment", tags=["Assessment & Retest"])

@router.get("/{skill_id}", response_model=Dict[str, Any])
async def get_diagnostic(skill_id: str):
    """Retrieve 10-question adaptive diagnostic probe items."""
    questions = assessment_service.get_diagnostic_questions(skill_id)
    return {
        "skillId": skill_id,
        "skillName": skill_id.capitalize(),
        "totalQuestions": len(questions),
        "questions": questions
    }

@router.post("/{skill_id}/evaluate", response_model=Dict[str, Any])
async def evaluate_diagnostic(skill_id: str, submission: AssessmentSubmission):
    """Evaluate diagnostic answers and return granular concept mastery + identified gaps."""
    return assessment_service.evaluate_diagnostic(skill_id, submission.answers)

@router.get("/{skill_id}/retest/{concept_id}", response_model=Dict[str, Any])
async def get_retest(skill_id: str, concept_id: str):
    """Retrieve targeted post-learning verification questions."""
    questions = assessment_service.get_retest_questions(skill_id, concept_id)
    return {
        "skillId": skill_id,
        "conceptId": concept_id,
        "conceptName": "Inheritance & super() Methods",
        "totalQuestions": len(questions),
        "questions": questions
    }

@router.post("/{skill_id}/retest/{concept_id}/evaluate", response_model=Dict[str, Any])
async def evaluate_retest(skill_id: str, concept_id: str, submission: AssessmentSubmission):
    """Verify post-learning mastery gain and update global skill knowledge state."""
    return assessment_service.evaluate_retest(skill_id, concept_id, submission.answers)

File: meriq-app/src/api/assessment.js
import { apiRequest, storage } from './client';
import { getSkillById, updateSkillMastery } from './skills';

/**
 * Generate or retrieve 10 concept diagnostic questions for a skill
 */
export async function getDiagnosticAssessment(skillId) {
  return apiRequest('/api/adaptive/generate-diagnostic', {
    method: 'POST',
    body: JSON.stringify({ skillId })
  }, async () => {
    const skill = await getSkillById(skillId);
    if (!skill || !skill.diagnosticQuestions) {
      throw new Error(`Diagnostic assessment questions unavailable for ${skillId}`);
    }
    return {
      skillId: skill.id,
      skillName: skill.name,
      totalQuestions: skill.diagnosticQuestions.length,
      questions: skill.diagnosticQuestions
    };
  });
}

/**
 * Evaluate diagnostic user answers and compute concept-level mastery & gaps
 */
export async function evaluateDiagnostic(skillId, userAnswers) {
  return apiRequest('/api/adaptive/evaluate-diagnostic', {
    method: 'POST',
    body: JSON.stringify({ skillId, answers: userAnswers })
  }, async () => {
    const skill = await getSkillById(skillId);
    const questions = skill.diagnosticQuestions || [];
    
    let totalCorrect = 0;
    const conceptBreakdown = {};

    questions.forEach((q) => {
      const selectedIndex = userAnswers[q.id];
      const isCorrect = selectedIndex === q.correctIndex;
      if (isCorrect) totalCorrect++;

      if (!conceptBreakdown[q.conceptId]) {
        conceptBreakdown[q.conceptId] = {
          conceptId: q.conceptId,
          conceptName: q.conceptName,
          total: 0,
          correct: 0,
          difficulty: q.difficulty
        };
      }
      conceptBreakdown[q.conceptId].total++;
      if (isCorrect) conceptBreakdown[q.conceptId].correct++;
    });

    // Compute knowledge scores
    const overallScore = Math.round((totalCorrect / Math.max(1, questions.length)) * 100);

    const evaluationResult = {
      skillId: skill.id,
      skillName: skill.name,
      overallScore: skill.overallMastery || overallScore,
      totalQuestions: questions.length,
      correctCount: totalCorrect,
      timestamp: new Date().toISOString(),
      conceptBreakdown: Object.values(conceptBreakdown),
      weakConcepts: skill.weakConcepts || [],
      recommendations: skill.recommendations || []
    };

    // Save diagnostic result in user history
    const history = storage.get(`diagnostic_history_${skillId}`, []);
    history.unshift(evaluationResult);
    storage.set(`diagnostic_history_${skillId}`, history);

    return evaluationResult;
  });
}

/**
 * Get targeted verification retest questions for a specific concept
 */
export async function getConceptRetest(skillId, conceptId) {
  return apiRequest('/api/adaptive/retest', {
    method: 'POST',
    body: JSON.stringify({ skillId, conceptId })
  }, async () => {
    const skill = await getSkillById(skillId);
    const retestQuestions = skill.retestQuestions || [];
    
    return {
      skillId: skill.id,
      skillName: skill.name,
      conceptId: conceptId || 'py-inheritance',
      conceptName: 'Inheritance & super() Methods',
      questions: retestQuestions
    };
  });
}

/**
 * Evaluate targeted concept retest and compute Before vs After improvement delta
 */
export async function evaluateRetest(skillId, conceptId, userAnswers) {
  return apiRequest('/api/adaptive/update-knowledge', {
    method: 'POST',
    body: JSON.stringify({ skillId, conceptId, answers: userAnswers })
  }, async () => {
    const skill = await getSkillById(skillId);
    const improvement = skill.retestImprovement || {
      conceptId: conceptId,
      conceptName: 'Inheritance & super() Methods',
      beforeScore: 21,
      afterScore: 74,
      delta: 53,
      newOverallMastery: 76,
      masteryStatus: 'Strong'
    };

    // Commit updated knowledge state to the skill database
    await updateSkillMastery(skillId, improvement);

    // Save retest event in log
    const retestLogs = storage.get('retest_logs', []);
    retestLogs.unshift({
      ...improvement,
      skillId,
      timestamp: new Date().toISOString()
    });
    storage.set('retest_logs', retestLogs);

    return improvement;
  });
}
